src/main/java/org/datadog/jenkins/plugins/datadog/listeners/DatadogGraphListener.java

```diff
@@ -0,0 +1,158 @@
+IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+THE SOFTWARE.
+ */
+
+package org.datadog.jenkins.plugins.datadog.listeners;
+
+import hudson.Extension;
+import hudson.model.Queue;
+import java.io.IOException;
+import java.util.Map;
+import java.util.Set;
+import java.util.logging.Logger;
+import javax.annotation.CheckForNull;
+import javax.annotation.Nonnull;
+import org.datadog.jenkins.plugins.datadog.DatadogClient;
+import org.datadog.jenkins.plugins.datadog.DatadogUtilities;
+import org.datadog.jenkins.plugins.datadog.clients.ClientFactory;
+import org.datadog.jenkins.plugins.datadog.model.BuildData;
+import org.datadog.jenkins.plugins.datadog.util.TagsUtil;
+import org.jenkinsci.plugins.workflow.actions.LabelAction;
+import org.jenkinsci.plugins.workflow.actions.StageAction;
+import org.jenkinsci.plugins.workflow.actions.ThreadNameAction;
+import org.jenkinsci.plugins.workflow.actions.TimingAction;
+import org.jenkinsci.plugins.workflow.cps.nodes.StepEndNode;
+import org.jenkinsci.plugins.workflow.cps.nodes.StepStartNode;
+import org.jenkinsci.plugins.workflow.flow.GraphListener;
+import org.jenkinsci.plugins.workflow.graph.BlockStartNode;
+import org.jenkinsci.plugins.workflow.graph.FlowNode;
+import org.jenkinsci.plugins.workflow.job.WorkflowRun;
+
+/**
+ * A GraphListener implementation which computes timing information
+ * for the various stages in a pipeline.
+ */
+@Extension
+public class DatadogGraphListener implements GraphListener {
+
+    private static final Logger logger = Logger.getLogger(DatadogGraphListener.class.getName());
+
+    @Override
+    public void onNewHead(FlowNode flowNode) {
+        if (!isMonitored(flowNode)) {
+            return;
+        }
+        StepEndNode endNode = (StepEndNode) flowNode;
+        StepStartNode startNode = endNode.getStartNode();
+        int stage_depth = 0;
+        for (BlockStartNode node : startNode.iterateEnclosingBlocks()) {
+            if (isStageNode(node)) {
+                stage_depth++;
+            }
+        }
+        DatadogClient client = ClientFactory.getClient();
+        try {
+            BuildData buildData = new BuildData(getRun(flowNode), flowNode.getExecution().getOwner().getListener());
+            String hostname = buildData.getHostname("");
+            Map<String, Set<String>> tags = buildData.getTags();
+            TagsUtil.addTagToTags(tags, "stage_name", getStageName(startNode));
+            TagsUtil.addTagToTags(tags, "stage_depth", String.valueOf(stage_depth));
+            tags.remove("result"); // Jenkins sometimes consider the build has a result even though it's still running.
```

Comment:
Do you know why this is? And why it's fine to remove for every metric?

Reply:
We're removing it only for the stageduration metrics which can never represent the status of a build. By definition the pipeline is still running so a result tag does not make sens.

Regarding why Jenkins does that I really don't know. I was consistently reproducing seeing a "result:SUCCESS" for just one stage in a big nested pipeline.

Reply:
Thanks for the explanation, makes sense!